### DISCUSSION TITLE
How to process Millions of files via Dynamic Task Mapping?

### DISCUSSION THREAD

**Discussion by vitaly-krugl** [apache/airflow#26172] [Q&A]
The "Dynamic Task Mapping" documentation (https://airflow.apache.org/docs/apache-airflow/2.3.0/concepts/dynamic-task-mapping.html) doesn't discuss the practical upper bound for the maximum number of dynamic tasks instances and the maximum list size that may be returned by an upstream task which generates the items for the dynamically mapped tasks.

It only discusses`[core] max_map_length` which allows to set the upper bound on the number of tasks that expand can create. However, this doesn't tell us anything about the practical side of things - what is the practical max for `[core] max_map_length`?

In my scenarios, my application needs to process millions of files from a networked drive. So, I would like to have my upstream task walk the filesystem and return a list containing millions of file paths for passing to the downstream dynamically mapped tasks for processing. This raises the following questions:

1. Is it practical for my upstream task to return a list containing millions of file paths? Would XCom and the rest of Airflow be able to handle it amicably? If not, what should it do instead?
2. Since millions of file paths can take up a lot of memory, is it possible/necessary for the upstream task to return a lazy list (or iterator) instead of an explicit list?
3. Will expand and airflow handle millions of dynamically mapped tasks amicably? If not, what should I do instead?

Thank you,
Vitaly

**Comment by potiuk** [comment]
Millions files to process in a single DAG is a lot and this is not what Airflow has been designed for and is probably bad idea in general to treat those files individually in any kind of system. Also likely your networked drive is not able to handle it, not Airflow - if you want to enumberate all the files in one job and distribute them to million of other jobs it will likely not work efficiently. The overhead of memory and CPU for starting a million python interpreters will be enormous. And no matter how much you distribute your work - if you will have one filesystem handling millions of files, likely your filesystem wll not handle it (unless it is totally distributed filesystem, well prepared for such a workload, effectively sharding and distributing storage and processing among multiple nodes - similar to  HDFS - but likely better - HDFS is designed to distribute parts of files among multiple nodes and depending on your file types it might be good or not). 

You should likely consider  rearchitecting your solution. Possibly you need a completely distributed filesystem that will be processing your files from multiple shards and batching you files and processing them with those shards independently from each other rather than centrialising them in a single DAG/job/Airflow instance. Airflow could orchestrate it (that's a good idea) but it should not process million files individually. Airlfow is not a "distributed processing" engine - it is orchestration engine, so you would likely use Airflow on "higher" layer of abstraction - defining and maanaging each shard separately, but then the job of processing each shard of files should be done by something that Airflow will orchestrate. And what that something should be depends very much on what kind of files you have, what you do with them and where you store the result.

GitHub discussion is likely not a very good place to discuss it.

Designing good architecture to perform effectively similar iknd of work requires paying someone a lot a architect to design it - based on multiple factors coming from many details of your system.
